gcsio/src/test/java/com/google/cloud/hadoop/gcsio/integration/GoogleCloudStorageGrpcIntegrationTest.java

```diff
@@ -0,0 +1,231 @@
+import com.google.cloud.hadoop.gcsio.GoogleCloudStorageItemInfo;
+import com.google.cloud.hadoop.gcsio.StorageResourceId;
+import com.google.cloud.hadoop.gcsio.ThrottledGoogleCloudStorage;
+import com.google.cloud.hadoop.gcsio.ThrottledGoogleCloudStorage.StorageOperation;
+import com.google.cloud.hadoop.gcsio.integration.GoogleCloudStorageTestHelper.TestBucketHelper;
+import com.google.common.base.Supplier;
+import com.google.common.base.Suppliers;
+import com.google.common.cache.CacheBuilder;
+import com.google.common.cache.CacheLoader;
+import com.google.common.cache.LoadingCache;
+import com.google.common.collect.Iterables;
+import com.google.common.util.concurrent.RateLimiter;
+import java.io.FileNotFoundException;
+import java.io.IOException;
+import java.nio.ByteBuffer;
+import java.nio.channels.ClosedChannelException;
+import java.nio.channels.SeekableByteChannel;
+import java.util.Arrays;
+import java.util.EnumSet;
+import org.junit.AfterClass;
+import org.junit.Before;
+import org.junit.Test;
+
+public class GoogleCloudStorageGrpcIntegrationTest {
+
+  // This prefix will be replaced by the prefix used in other gcs io integrate tests once it's
+  // whitelisted by GCS to access gRPC API.
+  private static final String BUCKET_NAME_PREFIX = "gcs-grpc-team-dataproc-it";
+
+  private static final Supplier<TestBucketHelper> BUCKET_HELPER =
+      Suppliers.memoize(() -> new TestBucketHelper(BUCKET_NAME_PREFIX));
+
+  private static final LoadingCache<GoogleCloudStorage, String> SHARED_BUCKETS =
+      CacheBuilder.newBuilder()
+          .build(
+              new CacheLoader<GoogleCloudStorage, String>() {
+                @Override
+                public String load(GoogleCloudStorage gcs) throws Exception {
+                  return BUCKET_HELPER.get().getUniqueBucketName("shared") + "_" + gcs.hashCode();
+                }
+              });
+
+  private final GoogleCloudStorage rawStorage;
+
+  public GoogleCloudStorageGrpcIntegrationTest() throws IOException {
+    GoogleCloudStorage gcs =
+        new ThrottledGoogleCloudStorage(
+            // Allow 2 create or delete bucket operation every second. This will hit rate limits,
+            // but GCS now has back-offs implemented for bucket operations.
+            RateLimiter.create(2),
+            new GoogleCloudStorageImpl(
+                GoogleCloudStorageTestHelper.getStandardOptionBuilder()
+                    .setGrpcEnabled(true)
+                    .build(),
+                GoogleCloudStorageTestHelper.getCredential()),
+            EnumSet.of(StorageOperation.DELETE_BUCKETS, StorageOperation.CREATE_BUCKET));
+    this.rawStorage = gcs;
+  }
+
+  @Before
+  public void setUp() throws IOException {
+    rawStorage.createBucket(getSharedBucketName());
+  }
+
+  @AfterClass
+  public static void cleanupBuckets() throws IOException {
+    BUCKET_HELPER.get().cleanup(Iterables.getLast(SHARED_BUCKETS.asMap().keySet()));
+  }
+
+  private String getSharedBucketName() {
```

Comment:
Why do we need this?

Reply:
We use this cache as a mean to persist a <storage, bucketName> so not to create random bucket name with helper class each time, plus we can use a trick to reference the storage object itself from cache in @Afterclass static method, but this name will not be accessible in @Beforeclass method if we want to reuse same bucket i.e. only create bucket once.

Refactored to just use original bucket prefix as common bucket name.